src/modules/fancyzones/lib/ZoneSet.cpp

```diff
@@ -162,41 +164,63 @@ IFACEMETHODIMP ZoneSet::AddZone(winrt::com_ptr<IZone> zone) noexcept
     return S_OK;
 }
 
-IFACEMETHODIMP_(winrt::com_ptr<IZone>)
-ZoneSet::ZoneFromPoint(POINT pt) noexcept
+IFACEMETHODIMP_(std::vector<int>)
+ZoneSet::ZonesFromPoint(POINT pt) noexcept
 {
-    winrt::com_ptr<IZone> smallestKnownZone = nullptr;
-    // To reduce redundant calculations, we will store the last known zones area.
-    int smallestKnownZoneArea = INT32_MAX;
-    for (auto iter = m_zones.rbegin(); iter != m_zones.rend(); iter++)
+    const int SENSITIVITY_RADIUS = 20;
+    std::vector<int> capturedZones;
+    for (size_t i = 0; i < m_zones.size(); i++)
+    {
+        auto zone = m_zones[i];
+        RECT newZoneRect = zone->GetZoneRect();
+        if (newZoneRect.left - SENSITIVITY_RADIUS <= pt.x && pt.x <= newZoneRect.right + SENSITIVITY_RADIUS &&
+            newZoneRect.top - SENSITIVITY_RADIUS <= pt.y && pt.y <= newZoneRect.bottom + SENSITIVITY_RADIUS &&
+            newZoneRect.left < newZoneRect.right && newZoneRect.top < newZoneRect.bottom) // proper zone
+        {
+            capturedZones.emplace_back(i);
+        }
+    }
+
+    // If captured zones do not overlap, return all of them
+    // Otherwise, return the smallest one
+
+    bool overlap = false;
+    for (size_t i = 0; i < capturedZones.size(); ++i)
     {
-        if (winrt::com_ptr<IZone> zone = iter->try_as<IZone>())
+        for (size_t j = i + 1; j < capturedZones.size(); ++j)
         {
-            RECT* newZoneRect = &zone->GetZoneRect();
-            if (PtInRect(newZoneRect, pt))
+            auto rectI = m_zones[capturedZones[i]]->GetZoneRect();
+            auto rectJ = m_zones[capturedZones[j]]->GetZoneRect();
+            if (max(rectI.top, rectJ.top) < min(rectI.bottom, rectJ.bottom) &&
+                max(rectI.left, rectJ.left) < min(rectI.right, rectJ.right))
```

Comment:
we should be able to check it w/o min/max, shouldn't we?
```suggestion
            if (rectI.left < rectJ.right && rectI.right > rectJ.left &&
                rectI.bottom > rectJ.top && rectI.top < rectJ.bottom)
```

Reply:
Yes of course, assuming there are no "improper" zones

Reply:
I'd suggest just `assert` on them at the begging